### PR TITLE
add CJ Bell's XML protocol docs

### PR DESCRIPTION
**Pull request by psteckler**
These are CJ Bell's XML protocol docs, which need some updating. But there is little other existing documentation, so these are a starting point.

This is basically the same file as his original document, at:
 https://github.com/siegebell/vscoq/wiki/XML-protocol

### CODE REVIEW

**Comment by ejgallego** [comment]
Cool. Maybe you would like to integrate my notes (cited in the document) too?

**Comment by ejgallego** [comment]
- https://github.com/ejgallego/jscoq/blob/master/etc/notes/coq-notes.md
- https://github.com/ejgallego/coq-serapi/blob/master/notes/async-support.md

**Comment by psteckler** [comment]
I'm thinking that those documents should be kept as separate pieces of text. Clearly, they're useful for someone wanting to use the XML protocol, but I'd say they're supportive, rather than integral to understanding the protocol. Also, they're more broadly useful than in the context of the protocol.

I could add them to my branch in /dev/doc/. 

There is some text that disclaims official Coq status, which I think you'd have to remove if we did that. :-)

**Comment by maximedenes** [comment]
@ejgallego what do you think? Should we merge this as it is or first integrate your notes?

**Comment by ejgallego** [comment]
I see no problem on merging as is, IIRC CJ notes already contained a link to my old notes. We should keep working on the docs indeed, for instance I'll do a pass once I get the XML thing ready.

**Comment by psteckler** [comment]
Should we add something to say that these notes are a draft and incomplete?

**Comment by psteckler** [comment]
One missing item I notice is the `About` call, which gives you the protocol version, *inter alia*.

In the document, it says `8.5` and `8.6`, and somewhere there could be a note explaining that those Coq versions correspond to particular protocol versions.

**Comment by ejgallego** [comment]
> In the document, it says 8.5 and 8.6, and somewhere there could be a note explaining that those Coq versions correspond to particular protocol versions.

Indeed good point. However I think in the long term the most reasonable thing is to have that file documenting only the current protocol, and rely on git to get the correct documentation for each version.

**Comment by psteckler** [comment]
> Indeed good point. However I think in the long term the most reasonable thing is to have that file documenting only the current protocol, and rely on git to get the correct documentation for each version.

I'm not sure about that. We haven't offered documentation for the protocol for 8.5 and 8.6. There could be people building clients that they'll want to use across multiple versions of the protocol, and they'll want to have all that information in one place, without having to fetch multiple versions of the document.

**Comment by ejgallego** [comment]
I see what you mean but note my "long term" qualification. I have to disagree on this one. When I open the documentation of glibc 2.7 I don't see stuff about 2.6, etc... That is what CHANGES are for.

IMO, the future should be:
- https://coq.inria.fr/coq/v8.6/xml_protocol.md
- https://coq.inria.fr/coq/v8.7/xml_protocol.md
- https://coq.inria.fr/coq/v8.8/xml_protocol.md

etc...

**Comment by ejgallego** [comment]
Thus I am fine with the current status, but I'd vote for getting rid of the compatibility stuff.

**Comment by psteckler** [comment]
Slicing the document that way is OK, I suppose, as long as all versions are easily available.

A difference with glibc is that one typically writes C code that works a particular version of it (or "version x and higher"), where a Coq client might want to work with many versions.

But not a big deal, as long as a coder can easily find the info.

**Comment by maximedenes** [comment]
So, should we merge as it is or slice the document now?

**Comment by ejgallego** [comment]
Merged or not, the documentation still needs lots of work, feel free to tweak as you see.

For instance, a "Known Problem" sections is mandatory, in particular due to the protocol having a few problem that won't be fixed soon (like the incorrect positioning bug)

**Comment by psteckler** [comment]
I'm going to cleanup the document a bit, maybe create the 8.5 and 8.6 versions.

Ideally, running `diff` on the different versions will show exactly the protocol differences.

**Comment by psteckler** [comment]
The entry for `feedback-ltacprof` looks to be bogus. When I tried 8.6, I got the Ltac profiling results table in a `feedback-custom`.

**Comment by psteckler** [comment]
I've split the document into 8.5 and 8.6 versions, did some cleanup/additions.

File: dev/doc/xml-protocol-8.5.md [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+for his [vscoq](https://github.com/siegebell/vscoq/) project.
+
+Here, the aim is to provide a "hands on" description of the XML
+protocol that coqtop and IDEs use to communicate. The protocol first appeared 
+with Coq 8.5, and is used by CoqIDE. It will also be used in upcoming 
+versions of Proof General.
+
+A somewhat out-of-date description of the async state machine is
+[documented here](https://github.com/ejgallego/jscoq/blob/master/etc/notes/coq-notes.md).
+OCaml types for the protocol can be [found here](https://github.com/coq/coq/blob/v8.5/ide/interface.mli).
```
Comment (Zimmi48) [inline dev/doc/xml-protocol-8.5.md]: you can replace this line by:

    OCaml types for the protocol can be found in the [`ide/interface.mli` file](/ide/interface.mli).

And the link will work on GitHub while not being dependent on this particular website (cf. example here https://github.com/Zimmi48/coq/blob/patch-2/dev/doc/xml-protocol-8.6.md).

File: dev/doc/xml-protocol-8.6.md [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+for his [vscoq](https://github.com/siegebell/vscoq/) project.
+
+Here, the aim is to provide a "hands on" description of the XML
+protocol that coqtop and IDEs use to communicate. The protocol first appeared 
+with Coq 8.5, and is used by CoqIDE. It will also be used in upcoming 
+versions of Proof General.
+
+A somewhat out-of-date description of the async state machine is
+[documented here](https://github.com/ejgallego/jscoq/blob/master/etc/notes/coq-notes.md).
+OCaml types for the protocol can be [found here](https://github.com/coq/coq/blob/v8.6/ide/interface.mli).
```
Comment (Zimmi48) [inline dev/doc/xml-protocol-8.6.md]: Same remark here.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
So now that you have done the split, why not submit two PRs, one on 8.5, one on 8.6? Even better, the 8.6 PR should contain both the 8.5 commit and a commit replacing the 8.5 version by the 8.6 version (so that one can easily find the diff e.g. by clicking on History in the GitHub interface).
Since GitHub is the preferred platform for looking at these Markdown files anyways, there is no advantage whatsoever in documenting the two versions of the protocol in the same branch since
https://github.com/coq/coq/blob/v8.5/dev/doc/xml-protocol.md and https://github.com/coq/coq/blob/v8.6/dev/doc/xml-protocol.md are as accessible as https://github.com/coq/coq/blob/trunk/dev/doc/xml-protocol-8.5.md and https://github.com/coq/coq/blob/trunk/dev/doc/xml-protocol-8.6.md.

**Comment by psteckler** [comment]
That sounds reasonable.

I think I will add a CHANGES section, so that implementors of multi-version IDEs will know where they have to handle differences in the protocol versions.

**Comment by psteckler** [comment]
Superseded by #562 and #563.

**Pull request closed without merging**